### PR TITLE
crash report fixes

### PR DESCRIPTION
**Pull request by pinzart90**
DYN-4849 - Handle exceptions coming from 3d party packages.

Fix sender object on Crash handlers
Only handle exceptions coming from assemblies that were loaded from a 3rd party package
Skip crash window and rethrow exceptions during tests
Removed feature flag for CER (will always be on)

### CODE REVIEW

File: src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs [hunk 1]
```diff
@@ -839,6 +839,18 @@ private void CrashGracefully(Exception ex, bool fatal = false)
         {
             try
             {
+                if (!fatal)
+                {
+                    var localPkg =  Model.GetPackageManagerExtension()?.PackageLoader?.LocalPackages?.FirstOrDefault(p => p.LoadedAssemblies.FirstOrDefault(a => a.Name == ex.Source) != null);
```
Comment (pinzart90) [inline src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs:844]: might happen that one of these package assemblies is already loaded by dynamo (as a dependency), so we could potentially continue running even though the exception is coming from somewhere else. (not the pacakge we identified)
But I think that might be an edge case.

Comment (pinzart90) [reply src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs:844]: At the moment this code will always continue the session and only log the exception. 
As an alternative, we could show a warning message that exception ex was thrown from package X and give the user a choice to continue the dynamo seasion or crash and send CER report 
@Amoursol any preferences?

**Comment by pinzart90** [comment]
Only the removePII fails (as in master) https://master-5.jenkins.autodesk.com/job/Dynamo/job/DynamoSelfServe/job/pullRequestValidation/15886/

File: src/DynamoCoreWpf/UI/Prompts/CrashPrompt.xaml.cs [hunk 2]
```diff
@@ -47,8 +48,14 @@ public CrashPrompt(DynamoViewModel dynamoViewModel)
             txtOverridingText.Text = string.Format(Wpf.Properties.Resources.CrashPromptDialogCrashMessage, productName);
         }
 
-        public CrashPrompt(CrashPromptArgs args, DynamoViewModel dynamoViewModel)
+        public CrashPrompt(CrashPromptArgs args, DynamoViewModel dynamoViewModel) : this(dynamoViewModel, args)
+        {}
+
+        internal CrashPrompt(object sender, CrashPromptArgs args)
         {
+            DynamoViewModel dynamoViewModel = sender as DynamoViewModel;
```
Comment (mjkkirschner) [inline src/DynamoCoreWpf/UI/Prompts/CrashPrompt.xaml.cs:56]: nit pick - use var

File: src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs (hunk 1 above)
Comment (mjkkirschner) [reply src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs:844]: `If the source property is not set explicitly, the runtime automatically sets it to the name of the assembly in which the exception originated.`

hmm, so in most cases this will work.

Have you verified that this property is set to the short name or full name of the assembly? Does it include the version number and culture?

File: src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs [hunk 3]
```diff
@@ -839,6 +839,18 @@ private void CrashGracefully(Exception ex, bool fatal = false)
...
+                    if (localPkg != null)
+                    {
+                        var crashDetails = new CrashErrorReportArgs(ex);
+                        Model?.Logger?.LogError($"Unhandled exception coming from package {localPkg.Name} was handled: {crashDetails.Details}");
```
Comment (mjkkirschner) [inline src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs:848]: any change that either of these properties are null?

**Comment by mjkkirschner** [comment]
can you add a test for this by raising a mocked exception and setting source explicitly to the name of an assembly in one of the packages we load by default during tests?

**Comment by pinzart90** [comment]
> can you add a test for this by raising a mocked exception and setting source explicitly to the name of an assembly in one of the packages we load by default during tests?

Sure can. Totally forgot about tests..

File: src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs (hunk 1 above)
Comment (mjkkirschner) [reply src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs:844]: what about the edge case where a user has shipped system binaries (Microsoft) or Dynamo binaries?

Comment (mjkkirschner) [reply src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs:844]: what about also logging something to ADP so we have analytics on how often these crashes occur without actually crashing?

File: src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs [hunk 4]
```diff
@@ -839,6 +839,18 @@ private void CrashGracefully(Exception ex, bool fatal = false)
...
+                        Analytics.TrackException(ex, true);
```
Comment (mjkkirschner) [inline src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs:849]: shouldn't fatal be set to false?

Comment (pinzart90) [reply src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs:849]: yep, forgot that. THanks !

File: src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs (hunk 1 above)
Comment (pinzart90) [reply src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs:844]: `Have you verified that this property is set to the short name or full name of the assembly`
It is just the short name. Obviously this can be overridden by custom exceptions.

Comment (pinzart90) [reply src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs:844]: `what about the edge case where a user has shipped system binaries (Microsoft) or Dynamo binaries?`
Not sure at the moment how I can identify these sort of cases...Are the package `LoadedAssemblies` actually loaded or they can just be on disk without actually being the ones to be loaded...I can check this

Comment (pinzart90) [reply src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs:844]: Yea, `LoadedAssemblies` could be already loaded assemblies. So if a user supplies a dotnet system dll, then all the exceptions from that dll (even if it was loaded by dotnet previously) will not cause a hard crash anymore.

I can verify if the assemblies have been loaded already, probably at a small performance penalty

File: src/DynamoCoreWpf/Utilities/CrashReportTool.cs [hunk 5]
```diff
@@ -183,14 +181,15 @@ internal static void ShowCrashWindow(object sender, CrashPromptArgs args)
         /// <param name="viewModel">The Dynamo view model</param>
```
Comment (aparajit-pratap) [inline src/DynamoCoreWpf/Utilities/CrashReportTool.cs:181]: `param name` needs to be updated.

**Review by aparajit-pratap — COMMENTED** [review]
Can you give some context to why was the crash happening in the first place?

File: src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs (hunk 1 above)
Comment (aparajit-pratap) [reply src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs:844]: Minor: does it make sense to have this check (code path) in the same function (`CrashGracefully`), as it isn't crashing in this case?

File: src/DynamoCoreWpf/UI/Prompts/CrashPrompt.xaml.cs (hunk 2 above)
Comment (pinzart90) [reply src/DynamoCoreWpf/UI/Prompts/CrashPrompt.xaml.cs:56]: you mean like so ? 
```
            DynamoModel model = null;
            var dynamoViewModel = sender as DynamoViewModel;
            if (dynamoViewModel != null)
            {
                model = dynamoViewModel.Model;
            }
            else if (sender is DynamoModel dm)
            {
                model = dm;
            }
```

**Comment by pinzart90** [comment]
> because an assembly that's already loaded is trying to be loaded again because it's a dependency of a

No, it is just a null reference exceptions or something similar

File: src/NodeServices/MessageEvents.cs [hunk 6]
```diff
@@ -66,9 +66,16 @@ internal static class DynamoConsoleLogger
     {
         public static event Action<string> LogMessageToDynamoConsole;
 
+        public static event Action<string> LogErrorToDynamoConsole;
```
Comment (pinzart90) [inline src/NodeServices/MessageEvents.cs:69]: Makes it easier to check logs from tests

File: src/DynamoPackages/Package.cs [hunk 7]
```diff
@@ -358,8 +361,14 @@ internal IEnumerable<PackageAssembly> EnumerateAndLoadAssembliesInBinDirectory()
             // Use the pkg header to determine which assemblies to load and prevent multiple enumeration
             // In earlier packages, this field could be null, which is correctly handled by IsNodeLibrary
             var nodeLibraries = Header.node_libraries;
+
+            List<string> appDomainAssemblies = null;
+            var binFiles = new DirectoryInfo(BinaryDirectory).EnumerateFiles("*.dll");
+            if (binFiles.Any()) {
+                appDomainAssemblies = AppDomain.CurrentDomain.GetAssemblies().Select(x => x.GetName().Name).ToList();
```
Comment (pinzart90) [inline src/DynamoPackages/Package.cs:368]: hopefully not too much of a performance hit here.
This will run for every loaded pkg that has at least one binary on disk.

Comment (pinzart90) [reply src/DynamoPackages/Package.cs:368]: If I try to move this a level up, the assembly list might change (because we iterate all the types and dependencies might get loaded)

File: src/DynamoCore/Models/DynamoModel.cs [hunk 8]
```diff
@@ -974,6 +974,7 @@ protected DynamoModel(IStartConfiguration config)
             LogWarningMessageEvents.LogWarningMessage += LogWarningMessage;
             LogWarningMessageEvents.LogInfoMessage += LogInfoMessage;
             DynamoConsoleLogger.LogMessageToDynamoConsole += LogMessageWrapper;
+            DynamoConsoleLogger.LogErrorToDynamoConsole += LogErrorMessageWrapper;
```
Comment (mjkkirschner) [inline src/DynamoCore/Models/DynamoModel.cs:977]: can this be unsubscribed somewhere?

Comment (pinzart90) [reply src/DynamoCore/Models/DynamoModel.cs:977]: done

File: src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs (hunk 1 above)
Comment (pinzart90) [reply src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs]: CrashGracefully is called from multiple places.
Another way I could go is to not subscribe to the unhandled_exception_handlers at all during testing.

**Review by aparajit-pratap — APPROVED** [review]

File: src/DynamoPackages/Package.cs (hunk 7 above)
Comment (mjkkirschner) [reply src/DynamoPackages/Package.cs:368]: have you considered how ALCs and packages containing extensions or binaries in other folders might affect this check?

Comment (mjkkirschner) [reply src/DynamoPackages/Package.cs:368]: Also I am still curious about the name property of the assembly, does that contain the version number or not?

Comment (pinzart90) [reply src/DynamoPackages/Package.cs]: I did not consider those. THanks !
` binaries in other folders might affect this check?` - CHanged the code to check if the faulty assembly is inside the package root folder.
`ALCs` - Changed the code to use TargetSite.Module.Assembly  - Gets the method(and assembly) where the exception was thrown - https://learn.microsoft.com/en-us/dotnet/api/system.exception.targetsite?view=net-8.0 (not sure why I did not see this before)
`curious about the name property of the assembly` - I am now using the equality operator on assemblies which should consider version and culture and ALCs

**Review by mjkkirschner — APPROVED** [review]

File: src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs [hunk 9]
```diff
@@ -848,10 +851,18 @@ private void CurrentDispatcher_UnhandledException(object sender, DispatcherUnhan
...
             }
 
-            // Try to handle the exception so that the host app can continue (in most cases).
-            // In some cases Dynamo code might still crash after this handler kicks in. In these edge cases
-            // we might see 2 CER windows (the extra one from the host app) - CER tool might handle this in the future.
-            e.Handled = true;
+            if (DynamoModel.IsTestMode)
+            {
+                // Do not handle the exception in test mode.
+                // Let the test host handle it.
```
Comment (aparajit-pratap) [inline src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs:857]: You might have covered this at standup but can you remind me again what you see in tests if the exception is marked as handled?

**Comment by aparajit-pratap** [comment]
@pinzart90 your new test is failing. Can you check? You might already be on top of it but I'm monitoring the builds this week ;)

File: src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs (hunk 9 above)
Comment (pinzart90) [reply src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs:857]: If (during tests) the exception is marked as handled inside the ex_handler, then the test that threw the exception might pass (as if nothing was thrown)
Our tests have a single entry point for async tasks (we run these with DispatcherUtil.DoLoop). Whenever an async task throws an exception (and if unhandled) it will propagate until the DIspatcherUtil.DoLoop calls and eventually be caught by the nunit test runner
